module/x/gravity/keeper/grpc_query.go

```diff
@@ -1,354 +1,407 @@
-// 		if err != nil {
-// 			panic("Invalid orchestrator addr in store!")
-// 		}
-// 		if reqOrchestrator.Equals(keyOrchestrator) {
-// 			return &types.QueryDelegateKeysByOrchestratorAddressResponse{ValidatorAddress: key.Validator, EthAddress: key.EthAddress}, nil
-// 		}
-
-// 	}
-// 	return nil, sdkerrors.Wrap(types.ErrInvalid, "No validator")
-// }
-
-// func (k Keeper) GetDelegateKeyByEth(
-// 	c context.Context,
-// 	req *types.QueryDelegateKeysByEthAddress) (*types.QueryDelegateKeysByEthAddressResponse, error) {
-// 	ctx := sdk.UnwrapSDKContext(c)
-// 	keys := k.GetDelegateKeys(ctx)
-// 	if err := types.ValidateEthereumAddress(req.EthAddress); err != nil {
-// 		return nil, sdkerrors.Wrap(err, "invalid eth address")
-// 	}
-// 	for _, key := range keys {
-// 		if req.EthAddress == key.EthAddress {
-// 			return &types.QueryDelegateKeysByEthAddressResponse{
-// 				ValidatorAddress:    key.Validator,
-// 				OrchestratorAddress: key.Orchestrator,
-// 			}, nil
-// 		}
-// 	}
-
-// 	return nil, sdkerrors.Wrap(types.ErrInvalid, "No validator")
-// }
-
-// func (k Keeper) GetPendingSendToEth(
-// 	c context.Context,
-// 	req *types.QueryPendingSendToEth) (*types.QueryPendingSendToEthResponse, error) {
-// 	ctx := sdk.UnwrapSDKContext(c)
-// 	batches := k.GetBatchTxes(ctx)
-// 	unbatchedTx := k.GetPoolTransactions(ctx)
-// 	senderAddress := req.SenderAddress
-// 	var res *types.QueryPendingSendToEthResponse
-
-// 	for _, batch := range batches {
-// 		for _, tx := range batch.Transactions {
-// 			if tx.Sender == senderAddress {
-// 				res.TransfersInBatches = append(res.TransfersInBatches, tx)
-// 			}
-// 		}
-// 	}
-
-// 	for _, tx := range unbatchedTx {
-// 		if tx.Sender == senderAddress {
-// 			res.UnbatchedTransfers = append(res.UnbatchedTransfers, tx)
-// 		}
-// 	}
-
-// 	return res, nil
-// }
+import (
+	"context"
+
+	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/gravity-bridge/module/x/gravity/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+)
+
+var _ types.QueryServer = Keeper{}
+
+func (k Keeper) Params(c context.Context, req *types.ParamsRequest) (*types.ParamsResponse, error) {
+	var params types.Params
+	k.paramSpace.GetParamSet(sdk.UnwrapSDKContext(c), &params)
+	return &types.ParamsResponse{Params: params}, nil
+}
+
+func (k Keeper) SignerSetTx(c context.Context, req *types.SignerSetTxRequest) (*types.SignerSetTxResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+
+	// TODO: audit once we finalize storage
+	var otx types.OutgoingTx
+
+	if req.Nonce == 0 {
+		store := prefix.NewStore(ctx.KVStore(k.storeKey), append([]byte{types.OutgoingTxKey}, types.SignerSetTxPrefixByte))
+		iter := store.ReverseIterator(nil, nil)
+		defer iter.Close()
```

Comment:
so this will return the latest stored value? what if there aren't any stored?

Reply:
@jackzampolin @jtremback is that possible?